src/main/java/com/nvidia/spark/rapids/jni/kudo/KudoTableMerger.java

```diff
@@ -155,80 +141,64 @@ private static int copyValidityBuffer(HostMemoryBuffer dest, int startBit,
-    int curDestByteIdx = startBit / 8;
-    int curDestBitIdx = startBit % 8;
-
-    while (curIdx < totalRowCount) {
-      int leftRowCount = totalRowCount - curIdx;
-      int appendCount;
-      if (curDestBitIdx == 0) {
-        appendCount = min(8, leftRowCount);
-      } else {
-        appendCount = min(8 - curDestBitIdx, leftRowCount);
-      }
-
-      int leftBitsInCurSrcByte = 8 - curSrcBitIdx;
-      byte srcByte = src.getByte(curSrcByteIdx);
-      if (leftBitsInCurSrcByte >= appendCount) {
-        // Extract appendCount bits from srcByte, starting from curSrcBitIdx
-        byte mask = (byte) (((1 << appendCount) - 1) & 0xFF);
-        srcByte = (byte) ((srcByte >>> curSrcBitIdx) & mask);
+    int totalRowCount = toIntExact(sliceInfo.getRowCount() + sliceInfo.getValidityBufferInfo().getBeginBit());
+    int curSrcIdx = sliceInfo.getValidityBufferInfo().getBeginBit();
+    int curDestIdx = startBit;
 
-        nullCount += (appendCount - NUMBER_OF_ONES[srcByte & 0xFF]);
 
-        // Sets the bits in destination buffer starting from curDestBitIdx to 0
-        byte destByte = dest.getByte(curDestByteIdx);
-        destByte = (byte) (destByte & ((1 << curDestBitIdx) - 1) & 0xFF);
+    while (curSrcIdx < totalRowCount) {
+      int leftRowCount = totalRowCount - curSrcIdx;
 
-        // Update destination byte with the bits from source byte
-        destByte = (byte) ((destByte | (srcByte << curDestBitIdx)) & 0xFF);
-        dest.setByte(curDestByteIdx, destByte);
+      int curDestOffset = (curDestIdx / 32) * Integer.BYTES;
+      int curDestBitIdx = curDestIdx % 32;
 
-        curSrcBitIdx += appendCount;
-        if (curSrcBitIdx == 8) {
-          curSrcBitIdx = 0;
-          curSrcByteIdx += 1;
-        }
-      } else {
-        // Extract appendCount bits from srcByte, starting from curSrcBitIdx
-        byte mask = (byte) (((1 << leftBitsInCurSrcByte) - 1) & 0xFF);
-        srcByte = (byte) ((srcByte >>> curSrcBitIdx) & mask);
-
-        byte nextSrcByte = src.getByte(curSrcByteIdx + 1);
-        byte nextSrcByteMask = (byte) ((1 << (appendCount - leftBitsInCurSrcByte)) - 1);
-        nextSrcByte = (byte) (nextSrcByte & nextSrcByteMask);
-        nextSrcByte = (byte) (nextSrcByte << leftBitsInCurSrcByte);
-        srcByte = (byte) (srcByte | nextSrcByte);
+      int curSrcOffset = srcOffset + (curSrcIdx / 32) * Integer.BYTES;
+      int curSrcBitIdx = curSrcIdx % 32;
 
-        nullCount += (appendCount - NUMBER_OF_ONES[srcByte & 0xFF]);
+      // This is safe since we always have validity buffer 4 bytes padded
+      int srcInt = src.getInt(curSrcOffset);
+      srcInt = srcInt >>> curSrcBitIdx;
 
-        // Sets the bits in destination buffer starting from curDestBitIdx to 0
-        byte destByte = dest.getByte(curDestByteIdx);
-        destByte = (byte) (destByte & ((1 << curDestBitIdx) - 1));
+      if (dest.getLength() >= (curDestOffset + Integer.BYTES)) {
+        // We have enough room to get an int
```

Comment:
Don't we _always_ have enough space to get an integer from the destination buffer because it's at least 4-byte padded?